### ISSUE TITLE
Deadlock during training on GPU backend, batch size dependent

### ISSUE THREAD

**Issue by lutogniew** [eclipse/deeplearning4j#8060]
Hi.

I have an issue when trying to train a CNN on GPU backend on Windows - the Java process seems to deadlock after the `Starting early stopping training` message, hanging indefinitely. No disk activity, no GPU activity, no CPU activity from the java PID. Everything works fine on CPU backend, though painfully slow - same command, same data, same network and code. 

I've posted this issue on Gitter, but got no response, hence this thread.

My problem seems similar to https://github.com/eclipse/deeplearning4j/issues/7637, but affects GPU and not CPU backend, and turning off MKL-DNN doesn't work in my case (or I did this incorrectly; details below).
However, setting batch size to 1 - or to any value up to and including 4 - works. No batch size above 4 that I've tried works.

I've successfully trained e.g. LSTMs on the same platform with GPU backend, with identical dependencies etc. - only different network and data. Trouble started when I tried using a custom CNN. I've also successfully run the `CnnSentenceClassificationExample.java` example on GPU backend, no problems.

Also, I've setup a `PerformanceListener` to print info every iteration. With CPU it starts outputting immediately, on GPU (when the deadlock manifests itself) it never prints anything - so it doesn't look like a single iteration is completed on GPU. Looks like Java stores some stuff in GPU memory (usage rises), then CPU cranks up for a moment, and then everything goes to idle. I've tried waiting for about 2h after that, nothing happened.
When it works, it starts outputting iterations almost immediately.

What's also interesting to me is that when the deadlock occurs, there is no data loaded into the Dl4j UI - no model visualization, no training overview, no system data ("Loading..." everywhere on the system page).

What I've tried:
* different minibatch sizes:
    * 1 - worked
    * 2 - worked
    * 3 - worked
    * 4 - worked
    * 5 - no
    * 8 - no
    * 16 - no
    * 30 - no
    * 32 - no
* disabling MKL-DNN
    * `Nd4jCuda.Environment.getInstance().setUseMKLDNN(false);` - doesn't work
    * `Nd4jCpu.Environment.getInstance().setUseMKLDNN(false);` - doesn't work
    * both set to false - doesn't work
* CPU backend - works every time

Do you have an idea as to what might be going on, or what I may try to fix/debug this?

Stack dump, system info, data shapes etc.: https://gist.github.com/lutogniew/f0ae24e3087f3c512c0080ace228f32b

Network configuration:
```
new NeuralNetConfiguration.Builder()
        .seed(123)
        .cudnnAlgoMode(ConvolutionLayer.AlgoMode.PREFER_FASTEST)
        .weightInit(WeightInit.RELU)
        .activation(Activation.SELU)
        .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
        .updater(new Adam(0.01))
        .convolutionMode(ConvolutionMode.Strict)
        .graphBuilder()
        .addInputs("input")
        .setInputTypes(InputType.convolutional(90, 68, 1))
        .addLayer("cnn3", new ConvolutionLayer.Builder()
                .kernelSize(3, 68)
                .stride(1, 68)
                .nOut(16)
                .build(), "input")
        .addLayer("pool", new Pooling2D.Builder(SubsamplingLayer.PoolingType.MAX)
                .kernelSize(2, 1)
                .padding(0, 0)
                .build(), "cnn3")
        .addLayer("out", new OutputLayer.Builder()
                .lossFunction(LossFunctions.LossFunction.MSE)
                .activation(Activation.IDENTITY)
                .nOut(1)
                .build(), "pool")
        .setOutputs("out")
        .build();
```

**Comment by raver119** [comment]
MKL-DNN isn't used for CUDA backend. cuDNN is used there, so i'd propose you to toggle it, and see what happens

**Comment by lutogniew** [comment]
Hey @raver119 , thanks for taking a moment to suggest something.

Unfortunately, that didn't help. I removed the `deeplearning4j-cuda-10.0` dependency and got the ` cuDNN not found: use cuDNN for better GPU performance` message, but differences ended there. It still deadlocked, and all other symtpoms remain. Still worked for `batchSize<=4`, still didn't for anything else.

I'll be grateful for any suggestions on how I could fix this, or on where I could to start debugging!

**Comment by raver119** [comment]
Wrap your iterator into AsyncShieldDataSetIterator.

```
val iterator = new AsyncShieldDataSetItterator(originalIterator);
```

and use this iterator for training.

**Comment by lutogniew** [comment]
@raver119 Yup, worked like magic, with `batchSize=128` and cuDNN enabled. Didn't finish training yet, however it trained two full epochs and continues. So, is there perhaps a flaw in my custom `RecordReader` that prevents async prefetch, or is this unrelated to my code? Is there anything I should fix, or anything I could improve?

One problem I see is that training seems extremely slow. With `batchSize=128` GPU usage is at around 0%, with CPU seemingly doing all the work. Performance is a bit disappointing:
```
2019-07-30 17:47:55,577 [main] - INFO t.p.m.ShapePrintingDataSetPreProcessor - Shapes:
    dataSet.getFeatures().shapeInfoToString(): Rank: 4, DataType: FLOAT, Offset: 0, Order: c, Shape: [95,1,90,68],  Stride: [6120,6120,68,1]
    dataSet.getLabels().shapeInfoToString(): Rank: 2, DataType: FLOAT, Offset: 0, Order: c, Shape: [95,1],  Stride: [1,1]
    dataSet.getFeaturesMaskArray().shapeInfoToString(): null
    dataSet.getLabelsMaskArray().shapeInfoToString(): null
2019-07-30 17:47:55,596 [main] - INFO o.d.o.listeners.PerformanceListener - ETL: 0 ms; iteration 7; iteration time: 38193 ms; samples/sec: 2.487; batches/sec: 0.026;
2019-07-30 17:47:55,596 [main] - INFO o.d.e.t.BaseEarlyStoppingTrainer - Completed training epoch 0
```

Is it because of the `AsyncShieldDataSetIterator`, or perhaps due to a small dataset...? I double checked that CUDA backend is used, with cuDNN enabled (no warnings about missing cuDNN).

**Comment by raver119** [comment]
Not quite sure. ASDI was introduce loooong ago, and that's the first deadlock complaint we've got so far.

cc @AlexDBlack

**Comment by lutogniew** [comment]
Not really complaining, I'm grateful for being able to use a great framework, esp. since it's open and free :) Thank you for helping so far, @raver119, and for the great work put into this framework.

I'm happy I didn't hit a dead-end. If I can help in any way, or if you find something new, I'll be happy to know. I'll try some tinkering on my end, in the mean time.

**Comment by raver119** [comment]
Yes you can help. Please share as much of your source code as you can: besides of NN arch i need to see early stopping trainer setup, need to know hardware details, and data transformation details.

**Comment by lutogniew** [comment]
Sure, I'll try to isolate a minimal example soon.

**Comment by AlexDBlack** [comment]
Yeah, we haven't had any issues with async iterator.
If async iterator was the cause here, I'd suspect a few possibilities:
1. An exception in the underlying iterator not propagated out
2. Something wrong with next/hasNext/reset in the underlying iterator: i.e., hasNext() returns true but next() doesn't have any data

One way to check/help isolate this: run the async iterator in a loop (no training), just calling .next() and .reset(), and see if  that deadlocks also.

**Comment by lutogniew** [comment]
@AlexDBlack Hey, thanks for the insights. I'll check that, esp. since I now think that the problem lies somewhere in the record reader-iterator part rather than actual training...

I just wanted to let you know that I didn't "abandon" this issue. While preparing the minimal example I noticed performance problems with my custom `RecordReader`, which might well be the cause of the perceived deadlock - so I just want to make absolutely sure that the `RecordReader` behaves correctly before I waste more of your time. I'm going to post something within a few days.

**Comment by AlexDBlack** [comment]
Sure, NP

This does give me an idea though - maybe we can add some sort of validation for this. (cc @raver119 )
Some sort of error like: 
```
Async iterator possible deadlock detected: underlying iterator .next() method has not returned after 60 seconds.
This may indicate a significant data loading bottleneck or a data pipeline bug.
```

**Comment by lutogniew** [comment]
I've run some experiments, and ever since I started pre-loading all the `DataSet`s into memory, I've encountered no problems - so I'm pretty sure the performance issue was due to the dismal performance of my custom `RecordReader`, and not a problem with DL4J. This is embarassing - sorry for wasting your time and for the delay in following up.

@AlexDBlack I think such a warning would be great. I'm all for verbose diagnostic messages where tricky problems might potentially occur.

So, I'm closing the issue. Thank you for your help!
